content/post/postgres/rfc-extension-packaging-lookup.md

```diff
@@ -0,0 +1,496 @@
+follow these steps:
+
+*   Set the `extension_path` GUC to search the site extension directory under
+    the packaging install. Something like
+    `$RPM_BUILD_ROOT/$(pg_config --extdir-vendor)`
+*   Install the extension into that directory:
+    `make install EXTDIR=$RPM_BUILD_ROOT`
+*   Run `make installcheck`
+
+This will allow PostgreSQL to find and load the extension during the tests.
+The Postgres installation will not have been modified; only the
+`extension_path` will have changed.
+
+### Docker/Kubernetes
+
+To allow extensions to be added to a Docker container and to persist beyond
+its lifetime, one or more [volumes] could be used. A couple of options:
+
+*   Mount the `--extdir-site` and/or `--extdir-vendor` directories as
+    persistent volumes (or one volume and a subdirectory for each). Then any
+    extensions installed into them will persist. Files for any one extension
+    will live on a single volume. If a new container spins up, as long as it
+    uses the same persistent volume(s), can access the same extensions.
+
+*   Create separate images for each extension, and then "install" them by
+    using the [Kubernetes image volume feature] to mount them as read-only
+    volumes in the appropriate subdirectory of `--extdir-site` or
+    `--extdir-vendor`. Thereafter, any new containers would simply have to
+    mount all the same extension image volumes persistently provide the same
+    extensions to all containers.
+
+### Postgres.app
+
+To allow extension installation without invalidating the Postgres.app bundle's
+signature, the app could be compiled to have `--extdir-site` and
+`--extdir-vendor` point to subdirectories well-known directories outside the
+app bundle, such as `/Library/Application Support/Postgres`.
+
+Any vendor or user extensions installed would be placed in those
+subdirectories, without changing the contents of the Postgres.app bundle.
+Postgres.app would know to find extensions in that location thanks to the
+inclusion of `$extdir_site` in the `extension_path` GUC.
+
+## Extension Directory Examples
+
+A core extension, like [citext], would live in
+`$(pg_config --extdir-core)/citext`, and have a structure such as:
+
+``` tree
+citext
+├── citext.control
+├── lib
+│   ├── citext.dylib
+│   └── bitcode
+│       ├── citext
+│       │   └── citext.bc
+│       └── citext.index.bc
+└── share
+    ├── citext--1.0--1.1.sql
+    ├── citext--1.1--1.2.sql
+    ├── citext--1.2--1.3.sql
+    ├── citext--1.3--1.4.sql
+    ├── citext--1.4--1.5.sql
+    ├── citext--1.4.sql
+    └── citext--1.5--1.6.sql
+```
+
+The subdirectory for a pure SQL extension named "pair" in a directory named
+“pair” that looks something like this:
+
+``` tree
+pair
+├── LICENSE.md
+├── README.md
+├── pair.control
+├── doc
+│   ├── html
+│   │   └── pair.html
+│   └── pair.md
+└── share
+    ├── pair--1.0--1.1.sql
+    └── pair--1.1.sql
+```
+
+A binary application like [pg_top] would live in the `pg_top` directory,
```

Comment:
Consistently with that I've said before, and despite understanding that PGXN does currently support not only "extensions" but also "packages", I disagree it's a good thing to distribute something like `pg_top` "as an extension". It is not, and IMO should not be put into the extensions directory.

Reply:
FWIW, the [PGXN Meta v2 RFC proposes](https://github.com/pgxn/rfcs/blob/64e6d63d3328b5cc012e2c980dd92572c5945992/text/0003-meta-spec-v2.md#contents) that its `contents` denotes different types of extensions: `CREATE EXTENSION` extensions, shared modules, and apps.

Reply:
I have no objection in packaging other tools "in a similar --or the same-- way" (e.g. as OCI images or whatever it takes). My objection is from considering them "like extensions" and living in the same directory (`extension_path`). I'd just place them anywhere else.

For the reasons I stated before, I'd not be happy placing file (including here binaries and potentially their shared libraries) in a place where Postgres would be looking for files to load into Postgres binary.

Besides, the UX for end users may not be good if now they need to either add multiple new entries to their `$PATH` variables or call these binaries with cumbersome paths as in `/usr/lib/postgresql/$major/extensions/pg_top/bin/pg_top` (vs simply `pg_top` if it is on a standard system binary path).

Reply:
Yeah, I have a comment in here about optionally symlinking bin files. But I need to think more about this.